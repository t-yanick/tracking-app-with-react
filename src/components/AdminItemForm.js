import React, { useState } from 'react';
import PropTypes from 'prop-types';

const AdminItemForm = ({
  title, unit, icon, target, handleSubmit,
}) => {
  const [error, setError] = useState('');
  const [inputTitle, setInputTitle] = useState(title || '');
  const [inputUnit, setInputUnit] = useState(unit || '');
  const [inputIcon, setInputIcon] = useState(icon || '');
  const [inputTarget, setInputTarget] = useState(target || '');

  const onTitleChange = (e) => {
    const { value } = e.target;
    const regex = /^[a-zA-Z0-9-':,()]{0,23}$/;
    if (value.match(regex)) {
      setError('');
      setInputTitle(value);
    } else {
      setError("Alphabet, numbers, ', ;, :, (, and ) are available. The number of letters should be within 23");
    }
  };

  const onUnitChange = (e) => {
    const { value } = e.target;
    const regex = /^[a-zA-Z]{0,9}$/;
    if (value.match(regex)) {
      setError('');
      setInputUnit(value);
    } else {
      setError('Alphabet is available and the number of letters should be between 1 and 9');
    }
  };

  const onIconChange = (e) => {
    const { value } = e.target;
    setInputIcon(value);
  };

  const onTargetChange = (e) => {
    const { value } = e.target;
    const regex = /^[0-9]{0,3}$/;
    if (value.match(regex)) {
      setError('');
      setInputTarget(value);
    } else {
      setError('Please provide a number with 3 digits');
    }
  };

  const onSubmit = (e) => {
    e.preventDefault();
    const iconRegex = /^[a-zA-Z0-9-]+:[a-zA-Z0-9-]+$/;
    if (!inputTitle || !inputUnit || !inputTarget) {
      setError('Please provide valid title, unit, and target');
    } else if (!!inputIcon && !inputIcon.match(iconRegex)) {
      setError('Please provide a valid icon string or leave it empty');
      setInputIcon('');
    } else {
      const item = {
        title: inputTitle,
        unit: inputUnit,
        icon: inputIcon,
        target: inputTarget,
      };
      handleSubmit(item);
    }
  };

  return (
    <div>
      <form className="form" onSubmit={onSubmit}>
        <div className="form__group">
          <div className="form__title">Item Label</div>
          <input
            type="text"
            name="title"
            placeholder="Item Label"
            value={inputTitle}
            onChange={onTitleChange}
          />
        </div>
        <div className="form__group">
          <div className="form__title">Unit (Singular)</div>
          <input
            type="text"
            name="unit"
            placeholder="Item Unit"
            value={inputUnit}
            onChange={onUnitChange}
          />
        </div>
        <div className="form__group">
          <div className="form__title">Target</div>
          <input
            type="text"
            name="target"
            placeholder="Item Target"
            value={inputTarget}
            onChange={onTargetChange}
          />
        </div>
        <div className="form__group">
          <div className="form__title">Icon</div>
          <p className="form__desc">
            Please add the string for data-icon of the icon that you can find in
            <span className="strong"><a href="https://iconify.design/icon-sets/" target="_blank" rel="noreferrer">Iconify</a></span>
            . For exapmle, if you want to use
            <span className="iconify" data-icon="bi:pen-fill" data-inline="false" />
            , you need to to add
            <span className="blue strong">&apos;bi:pen-fill&apos;</span>
            that is the data-icon attribute value of this icon. If you leave this field empty,
            <span className="iconify" data-icon="heroicons-outline:paper-clip" data-inline="false" />
            is provided automatically.
          </p>
          <input
            type="text"
            name="icon"
            placeholder="Item icon"
            value={inputIcon}
            onChange={onIconChange}
          />
        </div>
        {error && <p className="error-msg mb3">{error}</p>}
        <button type="submit" className="btn dark w100 mb2">Save Item</button>
      </form>
    </div>
  );
};

AdminItemForm.propTypes = {
  title: PropTypes.string,
  unit: PropTypes.string,
  icon: PropTypes.string,
  target: PropTypes.number,
  handleSubmit: PropTypes.func,
};

AdminItemForm.defaultProps = {
  title: '',
  unit: '',
  icon: '',
  target: 0,
  handleSubmit: null,
};

export default AdminItemForm;
