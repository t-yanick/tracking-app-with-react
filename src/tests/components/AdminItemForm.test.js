import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import AdminItemForm from '../../components/AdminItemForm';
import createTestStore from '../fixtures/createTestStore';

describe('AdminItemForm', () => {
  let store;
  beforeEach(() => {
    store = createTestStore();
  });
  test('should match with snapshot', () => {
    const handleSubmit = jest.fn();
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <AdminItemForm handleSubmit={handleSubmit} />
          </MemoryRouter>
        </Provider>,
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
