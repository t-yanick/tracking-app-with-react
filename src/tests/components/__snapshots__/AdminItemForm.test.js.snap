// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminItemForm should match with snapshot 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form__group"
    >
      <div
        className="form__title"
      >
        Item Label
      </div>
      <input
        name="title"
        onChange={[Function]}
        placeholder="Item Label"
        type="text"
        value=""
      />
    </div>
    <div
      className="form__group"
    >
      <div
        className="form__title"
      >
        Unit (Singular)
      </div>
      <input
        name="unit"
        onChange={[Function]}
        placeholder="Item Unit"
        type="text"
        value=""
      />
    </div>
    <div
      className="form__group"
    >
      <div
        className="form__title"
      >
        Target
      </div>
      <input
        name="target"
        onChange={[Function]}
        placeholder="Item Target"
        type="text"
        value=""
      />
    </div>
    <div
      className="form__group"
    >
      <div
        className="form__title"
      >
        Icon
      </div>
      <p
        className="form__desc"
      >
        Please add the string for data-icon of the icon that you can find in
        <span
          className="strong"
        >
          <a
            href="https://iconify.design/icon-sets/"
            rel="noreferrer"
            target="_blank"
          >
            Iconify
          </a>
        </span>
        . For exapmle, if you want to use
        <span
          className="iconify"
          data-icon="bi:pen-fill"
          data-inline="false"
        />
        , you need to to add
        <span
          className="blue strong"
        >
          'bi:pen-fill'
        </span>
        that is the data-icon attribute value of this icon. If you leave this field empty,
        <span
          className="iconify"
          data-icon="heroicons-outline:paper-clip"
          data-inline="false"
        />
        is provided automatically.
      </p>
      <input
        name="icon"
        onChange={[Function]}
        placeholder="Item icon"
        type="text"
        value=""
      />
    </div>
    
    <button
      className="btn dark w100 mb2"
      type="submit"
    >
      Save Item
    </button>
  </form>
</div>
`;
